weave/trace_server/calls_query_builder/object_ref_query_builder.py

```diff
@@ -0,0 +1,914 @@
+"""
+Object reference query builder for call queries.
+
+This module handles the specialized logic for filtering calls based on object references
+in expand columns. It provides functionality to:
+
+1. Detect object reference operands in queries
+2. Process object reference conditions into CTEs (Common Table Expressions)
+3. Generate optimized SQL for object reference filtering
+
+Key components:
+- ObjectRefCondition: Represents a condition that filters on object references
+- ObjectRefQueryProcessor: Handles processing of object ref conditions in queries
+- CTE building functions for efficient object reference filtering
+"""
+
+from abc import abstractmethod
+from typing import TYPE_CHECKING, Any, Optional, Union, get_args
+
+from pydantic import BaseModel
+
+from weave.trace_server.calls_query_builder.optimization_builder import (
+    QueryOptimizationProcessor,
+    apply_processor,
+)
+from weave.trace_server.calls_query_builder.utils import (
+    json_dump_field_as_sql,
+    param_slot,
+)
+from weave.trace_server.interface import query as tsi_query
+from weave.trace_server.orm import (
+    clickhouse_cast,
+    combine_conditions,
+    python_value_to_ch_type,
+    quote_json_path,
+)
+
+if TYPE_CHECKING:
+    from weave.trace_server.calls_query_builder.calls_query_builder import (
+        Condition,
+        OrderField,
+        ParamBuilder,
+    )
+
+
+# Add protection from ultra large queries, generally only realistically
+# constructed as an attack vector
+MAX_CTES_PER_QUERY = 10
+
+
+class ObjectRefCondition(BaseModel):
+    """Base class for object reference conditions"""
+
+    field_path: str
+    expand_columns: list[str]
+    case_insensitive: bool = False
+    conversion_type: Optional[tsi_query.CastTo] = None
+
+    @property
+    def is_table_rows_condition(self) -> bool:
+        # TODO: is this the only time this is an actual table join condition??
+        if (
```

Comment:
no! this is circumstantial it is perfectly possible for a table row to have a different path && for inputs.example to NOT be a table row